firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/serialization/FirestoreMapEncoder.kt

```diff
@@ -0,0 +1,208 @@
+    override fun encodeEnum(enumDescriptor: SerialDescriptor, index: Int) =
+        encodeValue(enumDescriptor.getElementName(index))
+
+    override val serializersModule: SerializersModule = EmptySerializersModule
+
+    override fun encodeNull(): Unit = encodedMap.let { it[Element(index++).encodeKey] = null }
+
+    // TODO: Handle @DocumentId and @ServerTimestamp annotations from descriptor
+    override fun encodeValue(value: Any): Unit =
+        encodedMap.let { it[Element(index++).encodeKey] = value }
+
+    override fun endStructure(descriptor: SerialDescriptor) = callback(encodedMap)
+
+    /**
+     * Recursively build the nested map when an encoded property has its own structure (i.e. a
+     * nested @[Serializable] object, or a nested list).
+     *
+     * @param descriptor the [SerialDescriptor] of the @[Serializable] object.
+     * @return a CompositeEncoder either to be a [FirestoreMapEncoder] or a [FirestoreListEncoder].
+     */
+    override fun beginStructure(descriptor: SerialDescriptor): CompositeEncoder {
+        // TODO: @DocumentID and @ServerTimeStamp should not be applied on Structures
+        if (depth == 0) {
+            return FirestoreMapEncoder(descriptor, depth = depth + 1) { encodedMap.putAll(it) }
+        }
+        val currentElement = Element(index++)
+        return when (descriptor.kind) {
+            StructureKind.CLASS -> {
+                FirestoreMapEncoder(descriptor, depth = depth + 1) {
+                    encodedMap[currentElement.encodeKey] = it
+                }
+            }
+            StructureKind.LIST -> {
+                FirestoreListEncoder(depth = depth + 1) {
+                    encodedMap[currentElement.encodeKey] = it
+                }
+            }
+            else -> {
+                throw IllegalArgumentException(
+                    "Incorrect format of nested object provided: <$descriptor.kind>"
+                )
+            }
+        }
+    }
+}
+
+/**
+ * Encodes a list of the @[Serializable] objects into a list of Firestore supported types.
+ *
+ * @param depth Object depth, defined as objects within objects.
+ * @param callback Puts encoded nested maps back to the parent map.
+ */
+private class FirestoreListEncoder(
+    private val depth: Int = 0,
+    private val callback: (MutableList<Any?>) -> Unit
+) : AbstractEncoder() {
+
+    /** A list that saves the encoding result. */
+    private val encodedList: MutableList<Any?> = mutableListOf()
+
+    /**
+     * Throw IllegalArgumentException if object depth, defined as objects within objects, larger
+     * than 500.
+     */
+    init {
+        if (depth == MAX_DEPTH) {
+            throw IllegalArgumentException(
+                "Exceeded maximum depth of $MAX_DEPTH, which likely indicates there's an object cycle"
+            )
+        }
+    }
+
+    override val serializersModule: SerializersModule = EmptySerializersModule
+
+    private var index: Int = 0
```

Comment:
do you need an index? from reading the code seems like `encondedList.size()` is the same?

Reply:
I have removed this index. Thank you.